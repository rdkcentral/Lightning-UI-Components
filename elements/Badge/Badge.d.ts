import type lng from '@lightningjs/core';
import type Base from '../../Base';
import type { Color, StylePartial } from '../../types/lui';
import type { TextBoxStyles } from '../TextBox';

export type BadgeStyles = {
  backgroundColor: Color;
  contentSpacing: number;
  iconColor: Color;
  paddingX: number;
  paddingY: number;
  radius: lng.Tools.CornerRadius;
  textStyle: TextBoxStyles;
  textAlign: string;
};

export default class Badge extends Base {
  title?: string;
  iconWidth?: number;
  iconColor?: Color;
  iconHeight?: number;
  icon?: string;
  iconAlign?: string;
  get style(): BadgeStyles;
  set style(v: StylePartial<BadgeStyles>);
}
