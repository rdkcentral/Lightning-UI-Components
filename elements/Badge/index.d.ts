import type Base from '../../Base';

export interface BadgeStyles {
  paddingX?: number;
  paddingY?: number;
  radius?: number;
  textStyle?: string;
  textAlign?: string;
  backgroundColor?: string;
  textColor?: string;
  contentSpacing?: number;
}

export default class Badge extends Base {
  title?: string;
  iconWidth?: number;
  iconColor?: string;
  iconHeight?: number;
  icon?: string;
  iconAlign?: string;
  styles?: BadgeStyles;
}
