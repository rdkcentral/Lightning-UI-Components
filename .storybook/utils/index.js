import React from 'react';

export const DocsLink = ({ children, id }) => {
  const docsmap = {
    lng: {
      _base: 'https://rdkcentral.github.io/Lightning/docs',
      Component: '/components/overview',
      Text: '/textures/text',
      Transition: '/transitions/attributes'
    }
  };
  const [source, path] = id.split('.');
  const href = docsmap[source]._base + docsmap[source][path];
  return (
    <a href={href} target="_blank">
      {children || id}
    </a>
  );
};

export const Item = ({ children }) => children;

/**
 * Creates formatted object for markdown prop tables
 * @param {Object} props
 * @param {Object} props.type - TS-style generic/utility type
 */
export const ObjectFormat = ({ object }) => {
  return [
    '{',
    <br />,
    '  ',
    ...Object.entries(object)
      .map(([k, v]) => [<span>&nbsp;&nbsp;</span>, `${k}: `, v, ',', <br />])
      .flat(),
    '}'
  ].map((x, i) => <Item key={i}>{x}</Item>);
};

/**
 * Formats TS-style generic types
 * @param {Object} props
 * @param {Object} props.children - component children
 * @param {String} props.type - Name of generic type
 */
export const GenericType = ({ children, type }) => {
  return [type, '<', children, '>'].map((x, i) => <Item key={i}>{x}</Item>);
};
