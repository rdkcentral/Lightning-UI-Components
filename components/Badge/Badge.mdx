import { Canvas, Story } from '@storybook/addon-docs/blocks';

# Badge

General purpose, Xfinity-branded badge

## Usage

The basic component just requires a title:

```js
import { Badge } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        title: 'HD'
      }
    };
  }
};
```

<Canvas>
  <Story id="badge--basic" />
</Canvas>

## API

### Properties

name|type|required|default|description
--|--|--|--|--
title|string|false|-|text to render
background|object|false|Xfinity theme|background style properties such as shader and color
textPosition|object|false|Xfinity theme|styling to apply to the text wrapper
textProperties|object|false|Xfinity theme|font properties to apply to the title
padding|number|false|8|padding to add on the left and right between the text and background
