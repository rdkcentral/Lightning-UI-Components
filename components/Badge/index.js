import lng from '@lightningjs/core';
import withStyles from '../../mixins/withStyles';
import Icon from '../Icon';

export const styles = theme => ({
  padding: theme.spacing(1),
  background: {
    h: theme.spacing(4),
    rect: true,
    shader: {
      type: lng.shaders.RoundedRectangle,
      radius: theme.border.radius.xsmall
    },
    color: theme.palette.badge.default
  },
  textPosition: {
    x: theme.spacing(1),
    y: 1
  },
  textProperties: {
    ...theme.typography.caption,
    color: theme.palette.text.light.primary,
    textAlign: 'center'
  },
  iconAlign: 'left'
});

class Badge extends lng.Component {
  static _template() {
    return {
      BadgeText: {},
      Icon: {
        type: Icon
      }
    };
  }

  _construct() {
    this._textPosition = this.styles.textPosition;
    this._textProperties = this.styles.textProperties;
    this._background = this.styles.background;
    this._padding = this.styles.padding;
    this._iconH = this._background.h - this._padding * 2;
    this.iconAlign = this.styles.iconAlign;
  }

  _init() {
    this._update();
  }

  _update() {
    this.patch({
      ...this._background,
      alpha: this._title || this._Icon ? 1 : 0
    });
    this._updateText();
    this._updateIcon();
  }

  _updateText() {
    this._BadgeText.on('txLoaded', () => {
      this._BadgeText.removeAllListeners();
      this._calculatePadding();
    });
    this._BadgeText.patch({
      ...this._textPosition,
      text: {
        ...this._textProperties,
        text: this._title || ''
      }
    });
  }

  _updateIcon() {
    this._Icon.on('txLoaded', () => {
      this._Icon.removeAllListeners();
      this._calculatePadding();
    });
    this._Icon.patch({
      icon: this._icon,
      w: this.iconWidth,
      h: this.iconHeight
    });
  }

  _calculateIconRatio() {
    if (this.iconWidth && this.iconHeight) {
      // use provided dimensions - required for SVGs
      return this.iconWidth / this.iconHeight;
    } else {
      // fallback to rendered dimensions - PNGs only
      return this._Icon.finalW / this._Icon.finalH;
    }
  }

  _calculatePadding() {
    this.h =
      this._background.h || this._BadgeText.renderHeight + this._padding * 2;

    const iconRatio = this._calculateIconRatio();

    // use iconRatio here to prevent a bug causing a PNG's ratio to be calculated incorrectly
    if (iconRatio) {
      // set icon dimensions
      this._Icon.h = this._iconH;
      this._Icon.w = this._iconH * iconRatio;
      this._Icon.y = (this.h - this._Icon.h) / 2;

      // set icon and text position
      if (this.iconAlign === 'left') {
        this._Icon.x = this._padding;
        this._BadgeText.x = this._Icon.x + this._Icon.w;
      } else if (this.iconAlign === 'right') {
        this._BadgeText.x = this._padding;
        this._Icon.x = this._BadgeText.x + this._BadgeText.renderWidth;
      }
    }

    this.w =
      this._BadgeText.renderWidth + this._padding * 2 + (this._Icon.w || 0);

    this.fireAncestors('$loadedBadge', this);
  }

  set title(title) {
    this._title = title;
    this._update();
  }

  get title() {
    return this._title;
  }

  set icon(icon) {
    this._icon = icon;
    this._update();
  }

  get icon() {
    return this._icon;
  }

  set iconHeight(iconHeight) {
    this._iconHeight = iconHeight;
    this._update();
  }

  get iconHeight() {
    return this._iconHeight;
  }

  set iconWidth(iconWidth) {
    this._iconWidth = iconWidth;
    this._update();
  }

  get iconWidth() {
    return this._iconWidth;
  }

  set background(background) {
    this._background = background;
    this._update();
  }

  get background() {
    return this._background;
  }

  set textPosition(textPosition) {
    this._textPosition = textPosition;
    this._update();
  }

  get textPosition() {
    return this._textPosition;
  }

  set textProperties(textProperties) {
    this._textProperties = textProperties;
    this._update();
  }

  get textProperties() {
    return this._textProperties;
  }

  set padding(padding) {
    this._padding = padding;
    this._update();
  }

  get padding() {
    return this._padding;
  }

  get _BadgeText() {
    return this.tag('BadgeText');
  }

  get _Icon() {
    return this.tag('Icon');
  }
}

export default withStyles(Badge, styles);
