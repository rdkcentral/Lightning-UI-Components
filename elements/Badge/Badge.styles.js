export const base = theme => ({
  contentSpacing: theme.spacer.sm,
  paddingX: theme.spacer.md,
  paddingY: theme.spacer.xs,
  radius: theme.radius.sm,
  textAlign: 'center',
  textStyle: theme.typography.caption1
});

export const palette = theme => ({
  neutral: {
    backgroundColor: theme.color.coreInverseSecondary,
    iconColor: theme.color.coreNeutral,
    textStyle: { textColor: theme.color.coreNeutral }
  },
  inverse: {
    backgroundColor: theme.color.coreNeutralSecondary,
    iconColor: theme.color.coreInverse,
    textStyle: { textColor: theme.color.coreInverse }
  },
  brand: {
    backgroundColor: theme.color.coreBrand,
    iconColor: theme.color.coreNeutral,
    textStyle: { textColor: theme.color.coreNeutral }
  }
});
