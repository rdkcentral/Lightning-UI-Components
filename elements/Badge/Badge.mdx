import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../.storybook/utils';
import Badge from '.';

# Badge

General purpose, Xfinity-branded badge

## Source

https://github.comcast.com/Lightning/lightning-ui/packages/@lightning/ui-core/src/components/Badge/Badge.js

## Usage

The basic component just requires a title:

```js
import { Badge } from '@lightning/ui';
import circle from 'assets/images/circle.svg';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        title: 'HD'
      }
    };
  }
}
```

<Canvas>
  <Story id="elements-badge--text" />
</Canvas>

Or an icon:

```js
import { Badge } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        icon: circle,
        iconWidth: 50,
        iconHeight: 50
      }
    };
  }
}
```

<Canvas>
  <Story id="elements-badge--icon" />
</Canvas>

A title and an icon can also be used together:

```js
import { Badge } from '@lightning/ui';
import lightning from 'assets/images/lightning.png';

class Example extends lng.Component {
  static _template() {
    return {
      Badge: {
        type: Badge,
        title: 'HD',
        icon: lightning
      }
    };
  }
}
```

<Canvas>
  <Story id="elements-badge--text-with-icon" />
</Canvas>

## API

### Properties

| name      | type   | required | default | description                                                          |
| --------- | ------ | -------- | ------- | -------------------------------------------------------------------- |
| title     | string | false    | -       | text to render                                                       |
| icon      | string | false    | -       | path to image or inline SVG XML                                      |
| padding   | number | false    | -       | padding to add on the left and right between the text and background |
| iconWidth | string | false    | -       | icon width (required for SVG icons)                                  |
| iconColor | string | false    | -       | color to be applied to the icon (only works with pure white PNGs)    |
| iconAlign | string | false    | -       | side of the text the icon will appear on (`left` or `right`)         |

### Style Properties

| name            | type          | description                                                          |
| --------------- | ------------- | -------------------------------------------------------------------- |
| backgroundColor | string        | background style properties such as shader and color                 |
| contentSpacing  | number        | padding to add between the icon and the text                         |
| iconColor       | string        | color of the icon next to the text                                   |
| paddingX        | number        | padding to add on the left and right between the text and background |
| paddingY        | number        | padding to add on the top and bottom between the text and background |
| radius          | number\|array | shape of component                                                   |
| textAlign       | string        | side of the component the text will appear on (`left` or `right`)    |
| textStyle       | string        | font properties to apply to the title                                |

### Events

#### $loadedBadge():void

Fires when the contents of the badge have finished loading and the background has been appropriately sized based on the text input.
