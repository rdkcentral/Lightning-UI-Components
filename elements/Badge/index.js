import lng from '@lightningjs/core';
import Base from '../../Base';
import Icon from '../Icon';
import * as styles from './Badge.styles';
import TextBox from '../TextBox';
import { withExtensions } from '../../mixins';
import { getHexColor } from '../../utils';

class Badge extends Base {
  static _template() {
    return {
      rect: true,
      shader: {
        type: lng.shaders.RoundedRectangle
      },
      BadgeText: {
        type: TextBox,
        mountY: 0.5,
        signals: {
          textBoxChanged: '_badgeTextLoaded'
        }
      },
      Icon: {
        type: Icon,
        mountY: 0.5,
        signals: {
          itemChanged: '_iconLoaded'
        }
      }
    };
  }

  static get __componentName() {
    return 'Badge';
  }

  static get __themeStyles() {
    return styles;
  }

  static get properties() {
    return [
      'icon',
      'title',
      'iconWidth',
      'iconColor',
      'iconHeight',
      'iconAlign'
    ];
  }

  static get tags() {
    return ['BadgeText', 'Icon'];
  }

  _setTitle(v) {
    this._badgeTextPromise = new Promise(resolve => {
      this._badgeTextPromiseResolver = resolve;
    });
    return v;
  }

  _setIcon(v) {
    this._iconPromise = new Promise(resolve => {
      this._iconPromiseResolver = resolve;
    });
    return v;
  }

  _setIconWidth(v) {
    this._iconPromise = new Promise(resolve => {
      this._iconPromiseResolver = resolve;
    });
    return v;
  }

  _setIconHeight(v) {
    this._iconPromise = new Promise(resolve => {
      this._iconPromiseResolver = resolve;
    });
    return v;
  }

  async _update() {
    this._updateText();
    this._updateIcon();
    await Promise.all(
      [this._badgeTextPromise, this._iconPromise].filter(Boolean)
    );
    this._updateWidth();
    this._updateBackground();
    this._updatePositions();
  }

  _badgeTextLoaded() {
    this._badgeTextPromiseResolver && this._badgeTextPromiseResolver();
  }

  _iconLoaded() {
    this._iconPromiseResolver && this._iconPromiseResolver();
  }

  _updateBackground() {
    this.patch({
      h: Math.max(this._BadgeText.h, this._Icon.h) + this.style.paddingY * 2,
      color: this.style.backgroundColor,
      shader: { radius: this.style.radius }
    });
  }

  _updateText() {
    if (this._BadgeText) {
      this._BadgeText.patch({
        textAlign: this.style.textAlign,
        textStyle: this.style.textStyle,
        content: this.title || '',
        x: this.style.paddingX,
        y: this._h / 2
      });
    }
  }

  _updateIcon() {
    this._Icon.patch({
      icon: this.icon,
      w: this.iconWidth,
      h: this.iconHeight,
      style: {
        color: getHexColor(this.style.iconColor)
      }
    });
  }

  _updateWidth() {
    let contentSpacing = 0;
    if (this.icon && this.title) {
      contentSpacing = this.style.contentSpacing;
    }
    this.w = this.title
      ? this._BadgeText.renderWidth +
        this.style.paddingX * 2 +
        (this._Icon.finalW || 0) +
        contentSpacing
      : this.style.paddingX * 2 + (this._Icon.finalW || 0);
  }

  _updatePositions() {
    this._Icon.y = this.h / 2;
    // set icon and text position
    if (this.iconAlign === 'left' && this.title) {
      this._Icon.x = this.style.paddingX;
      this._BadgeText.x =
        this._Icon.x + this._Icon.finalW + this.style.contentSpacing;
    } else if (this.iconAlign === 'right' && this.title) {
      this._BadgeText.x = this.style.paddingX;
      this._Icon.x =
        this._BadgeText.x +
        this._BadgeText.renderWidth +
        this.style.contentSpacing;
    } else {
      this._Icon.x = this.style.paddingX;
    }
    this.fireAncestors('$loadedBadge', this);
    this._BadgeText.y = this.h / 2; // Set new alignment for badge text
  }

  _getIconHeight() {
    if (this.icon) {
      if (!this._Icon.finalH) {
        if (this._BadgeText._Text) {
          return this._BadgeText._Text.text.lineHeight;
        }
      } else {
        return this._Icon.finalH;
      }
    } else {
      return 0;
    }
  }
}

export default withExtensions(Badge);
