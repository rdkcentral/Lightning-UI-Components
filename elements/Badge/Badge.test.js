import TestUtils from '../../test/lightning-test-utils';
import Badge from '.';
import { getHexColor } from '../../utils';

const createBadge = TestUtils.makeCreateComponent(Badge);

const circleIcon = TestUtils.pathToDataURI('assets/images/circle.svg');
const xfinityLogo = TestUtils.pathToDataURI(
  'assets/images/Xfinity-Provider-Logo-Square.png'
);

describe('Badge', () => {
  let badge, testRenderer;

  beforeEach(() => {
    [badge, testRenderer] = createBadge();
  });
  afterEach(() => {
    badge = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the badge text', () => {
    const title = 'HD';
    badge.title = title;
    testRenderer.forceAllUpdates();
    expect(badge.announce).toBe(title);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    badge.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(badge.announce).toBe(overrideString);
  });

  it('renders icon and text side by side', () => {
    // eslint-disable-next-line no-unused-vars
    const [_, testRenderer] = createBadge({
      title: 'HD',
      icon: circleIcon
    });
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('renders icon only', () => {
    // eslint-disable-next-line no-unused-vars
    const [_, testRenderer] = createBadge({
      icon: circleIcon
    });
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('should update the title', async done => {
    const title = 'HD';
    [badge, testRenderer] = createBadge(
      { title },
      { spyOnMethods: ['_update', '_updateBackground'] }
    );
    setTimeout(() => {
      badge._badgeTextLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    expect(badge.title).toBe(title);
    expect(badge.w).toBe(18);
    const title2 = 'longer title';
    badge.title = title2;
    setTimeout(() => {
      badge._badgeTextLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    expect(badge.title).toBe(title2);
    expect(badge.w).toBe(28);
    expect(badge.h).toBe(32);
    done();
  });

  it('should update the background', () => {
    const color = getHexColor('ffffff', 0.25);
    [badge] = createBadge({ background: { color }, icon: undefined });
    expect(badge.background.color).toBe(color);
  });

  it('should update the text style', async done => {
    const color = getHexColor('ffffff', 0.25);
    const title = 'HD';
    const icon = xfinityLogo;
    const textProperties = { color, textAlign: 'left' };

    [badge, testRenderer] = createBadge(
      {
        title,
        icon,
        h: 30,
        iconAlign: 'right',
        iconWidth: 50,
        iconHeight: 50
      },
      { spyOnMethods: ['_updateBackground'] }
    );
    badge.style.paddingX = 0;
    setTimeout(() => {
      badge._badgeTextLoaded();
      badge._iconLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    badge._BadgeText.text = textProperties;
    expect(badge._BadgeText.x).toBe(0);
    expect(badge._BadgeText.y).toBe(badge.h / 2);
    // This should work once bug in TextBox is fixed to patch textStyle
    // expect(badge._BadgeText._Text.text.textAlign).toBe(
    //   textProperties.textAlign
    // );
    //expect(badge._BadgeText._Text.text.color).toBe(color);
    done();
  });

  it('should update the icon', () => {
    const icon = xfinityLogo;
    [badge, testRenderer] = createBadge({ icon });
    expect(badge.icon).toBe(icon);
    const icon2 = circleIcon;
    badge.icon = icon2;
    expect(badge.icon).toBe(icon2);
    expect(badge._Icon.color).toBe(badge._BadgeText.color);
  });

  it('should position the text and icon based on the iconAlign property', async done => {
    const title = 'HD';
    const icon = xfinityLogo;
    [badge, testRenderer] = createBadge(
      {
        title,
        icon,
        iconAlign: 'right',
        iconWidth: 50,
        iconHeight: 50
      },
      {
        spyOnMethods: ['_updateBackground', '_updateIcon']
      }
    );
    setTimeout(() => {
      badge._badgeTextLoaded();
      badge._iconLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;

    expect(badge._BadgeText.x).toEqual(badge.style.paddingX);
    expect(badge._Icon.x).toEqual(
      badge._BadgeText.x +
        badge._BadgeText.renderWidth +
        badge.style.contentSpacing
    );
    done();
  });

  it('should position the text and icon based on the iconAlign property', async done => {
    const title = 'HD';
    const icon = xfinityLogo;
    [badge, testRenderer] = createBadge(
      {
        title,
        icon,
        iconAlign: 'left',
        iconWidth: 50,
        iconHeight: 50
      },
      {
        spyOnMethods: ['_updateBackground', '_updateIcon']
      }
    );
    setTimeout(() => {
      badge._badgeTextLoaded();
      badge._iconLoaded();
    }, 500);
    await badge.__updateBackgroundSpyPromise;
    expect(badge._Icon.x).toEqual(8);
    done();
  });
});
