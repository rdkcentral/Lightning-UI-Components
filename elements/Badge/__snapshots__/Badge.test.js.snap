// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders 1`] = `
Object {
  "Component": Object {
    "BadgeText": Object {
      "active": false,
      "alpha": 0.001,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "BadgeText",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "BadgeText",
      ],
      "type": "TextBox",
      "visible": true,
      "w": 0,
      "x": 8,
      "y": 4,
      "zIndex": 0,
    },
    "Icon": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Icon",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "Icon",
      ],
      "type": "Icon",
      "visible": true,
      "w": 0,
      "x": 8,
      "y": 4,
      "zIndex": 0,
    },
  },
}
`;

exports[`Badge renders icon and text side by side 1`] = `
Object {
  "Component": Object {
    "BadgeText": Object {
      "active": false,
      "alpha": 0.001,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Text": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Text",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Text",
          ],
          "texture": Object {
            "fontFace": "DMSansBold",
            "fontSize": 20,
            "lineHeight": 24,
            "text": "HD",
            "textAlign": "center",
            "textBaseline": "bottom",
            "textColor": 4294375161,
            "type": "TextTexture",
            "verticalAlign": "middle",
          },
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 2,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "BadgeText",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "BadgeText",
      ],
      "type": "TextBox",
      "visible": true,
      "w": 0,
      "x": 8,
      "y": 0,
      "zIndex": 0,
    },
    "Icon": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Icon",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "Icon",
      ],
      "texture": Object {
        "src": "data:image/jpg;base64, PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGhlaWdodD0iMTAwIiB3aWR0aD0iMTAwIj48Y2lyY2xlIGN4PSI1MCIgY3k9IjUwIiByPSI0MCIgc3Ryb2tlPSJibGFjayIgc3Ryb2tlLXdpZHRoPSIzIiBmaWxsPSJyZWQiIC8+PC9zdmc+",
        "type": "ImageTexture",
      },
      "type": "Icon",
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;

exports[`Badge renders icon only 1`] = `
Object {
  "Component": Object {
    "BadgeText": Object {
      "active": false,
      "alpha": 0.001,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "BadgeText",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "BadgeText",
      ],
      "type": "TextBox",
      "visible": true,
      "w": 0,
      "x": 8,
      "y": 0,
      "zIndex": 0,
    },
    "Icon": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Icon",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "Icon",
      ],
      "texture": Object {
        "src": "data:image/jpg;base64, PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIGhlaWdodD0iMTAwIiB3aWR0aD0iMTAwIj48Y2lyY2xlIGN4PSI1MCIgY3k9IjUwIiByPSI0MCIgc3Ryb2tlPSJibGFjayIgc3Ryb2tlLXdpZHRoPSIzIiBmaWxsPSJyZWQiIC8+PC9zdmc+",
        "type": "ImageTexture",
      },
      "type": "Icon",
      "visible": true,
      "w": 0,
      "x": 0,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
